diff --git a/Makefile b/Makefile
index c2a0b33..bdf7905 100644
--- a/Makefile
+++ b/Makefile
@@ -61,6 +61,8 @@ SAIL_DEFAULT_INST += riscv_insts_vext_fp_red.sail
 SAIL_DEFAULT_INST += riscv_insts_zicbom.sail
 SAIL_DEFAULT_INST += riscv_insts_zicboz.sail
 
+SAIL_DEFAULT_INST += riscv_insts_custom.sail
+
 SAIL_SEQ_INST  = $(SAIL_DEFAULT_INST) riscv_jalr_seq.sail
 SAIL_RMEM_INST = $(SAIL_DEFAULT_INST) riscv_jalr_rmem.sail riscv_insts_rmem.sail
 
diff --git a/custom_inst_test/min3_binutils_test.S b/custom_inst_test/min3_binutils_test.S
new file mode 100644
index 0000000..338262d
--- /dev/null
+++ b/custom_inst_test/min3_binutils_test.S
@@ -0,0 +1,35 @@
+.section .text.init
+.globl _start
+
+_start:
+    li x6, 10
+    li x7, 3
+
+    # funct7  rs2   rs1   funct3 rd    opcode  
+    # 0000000 00111 00110 000    00101 0001011
+    min3 x5, x6, x7
+
+    li x10, 3
+    beq x5, x10, pass
+
+fail:
+    la t1, tohost
+    li t0, 3
+    sd t0, 0(t1)
+    j fail
+
+pass:
+    la t1, tohost
+    li t0, 1
+    sd t0, 0(t1)
+    j pass
+
+.section .tohost, "aw", @progbits
+.align 3
+.globl tohost
+tohost:
+    .dword 0
+
+.globl fromhost
+fromhost:
+    .dword 0
diff --git a/custom_inst_test/min3_binutils_test.dump b/custom_inst_test/min3_binutils_test.dump
new file mode 100644
index 0000000..90bb092
--- /dev/null
+++ b/custom_inst_test/min3_binutils_test.dump
@@ -0,0 +1,26 @@
+
+min3_binutils_test.elf:     file format elf64-littleriscv
+
+
+Disassembly of section .text.init:
+
+0000000080000000 <_start>:
+    80000000:	4329                	li	t1,10
+    80000002:	438d                	li	t2,3
+    80000004:	0073028b          	min3	t0,t1,t2
+    80000008:	450d                	li	a0,3
+    8000000a:	00a28a63          	beq	t0,a0,8000001e <pass>
+
+000000008000000e <fail>:
+    8000000e:	00001317          	auipc	t1,0x1
+    80000012:	ff230313          	addi	t1,t1,-14 # 80001000 <tohost>
+    80000016:	428d                	li	t0,3
+    80000018:	00533023          	sd	t0,0(t1)
+    8000001c:	bfcd                	j	8000000e <fail>
+
+000000008000001e <pass>:
+    8000001e:	00001317          	auipc	t1,0x1
+    80000022:	fe230313          	addi	t1,t1,-30 # 80001000 <tohost>
+    80000026:	4285                	li	t0,1
+    80000028:	00533023          	sd	t0,0(t1)
+    8000002c:	bfcd                	j	8000001e <pass>
diff --git a/custom_inst_test/min3_binutils_test.elf b/custom_inst_test/min3_binutils_test.elf
new file mode 100755
index 0000000..d5d7055
Binary files /dev/null and b/custom_inst_test/min3_binutils_test.elf differ
diff --git a/custom_inst_test/min3_test.S b/custom_inst_test/min3_test.S
new file mode 100644
index 0000000..8e70a6a
--- /dev/null
+++ b/custom_inst_test/min3_test.S
@@ -0,0 +1,35 @@
+.section .text.init
+.globl _start
+
+_start:
+    li x6, 10
+    li x7, 3
+
+    # funct7  rs2   rs1   funct3 rd    opcode  
+    # 0000000 00111 00110 000    00101 0001011
+    .word 0x0073028b    # min3 x5, x6, x7
+
+    li x10, 3
+    beq x5, x10, pass
+
+fail:
+    la t1, tohost
+    li t0, 3
+    sd t0, 0(t1)
+    j fail
+
+pass:
+    la t1, tohost
+    li t0, 1
+    sd t0, 0(t1)
+    j pass
+
+.section .tohost, "aw", @progbits
+.align 3
+.globl tohost
+tohost:
+    .dword 0
+
+.globl fromhost
+fromhost:
+    .dword 0
diff --git a/custom_inst_test/min3_test.dump b/custom_inst_test/min3_test.dump
new file mode 100644
index 0000000..b9882a2
--- /dev/null
+++ b/custom_inst_test/min3_test.dump
@@ -0,0 +1,24 @@
+
+min3_test.elf:     file format elf64-littleriscv
+
+
+Disassembly of section .text:
+
+0000000080000000 <_start>:
+    80000000:	4329                	li	t1,10
+    80000002:	438d                	li	t2,3
+    80000004:	0073028b          	.word	0x0073028b
+    80000008:	450d                	li	a0,3
+    8000000a:	00a28863          	beq	t0,a0,8000001a <pass>
+
+000000008000000e <fail>:
+    8000000e:	00000313          	li	t1,0
+    80000012:	428d                	li	t0,3
+    80000014:	00533023          	sd	t0,0(t1)
+    80000018:	bfdd                	j	8000000e <fail>
+
+000000008000001a <pass>:
+    8000001a:	00000313          	li	t1,0
+    8000001e:	4285                	li	t0,1
+    80000020:	00533023          	sd	t0,0(t1)
+    80000024:	bfdd                	j	8000001a <pass>
diff --git a/custom_inst_test/min3_test.elf b/custom_inst_test/min3_test.elf
new file mode 100755
index 0000000..8873860
Binary files /dev/null and b/custom_inst_test/min3_test.elf differ
diff --git a/custom_inst_test/min3_test.ld b/custom_inst_test/min3_test.ld
new file mode 100644
index 0000000..0595f62
--- /dev/null
+++ b/custom_inst_test/min3_test.ld
@@ -0,0 +1,27 @@
+OUTPUT_ARCH(riscv)
+ENTRY(_start)
+
+SECTIONS
+{
+  . = 0x80000000;
+
+  .text.init : {
+    *(.text.init)
+    *(.text*)
+  }
+
+  . = ALIGN(0x1000);
+
+  .tohost : {
+    *(.tohost)
+  }
+
+  .data : {
+    *(.data*)
+  }
+
+  .bss : {
+    *(.bss*)
+    *(COMMON)
+  }
+}
\ No newline at end of file
diff --git a/model/riscv_insts_custom.sail b/model/riscv_insts_custom.sail
new file mode 100644
index 0000000..ace03ee
--- /dev/null
+++ b/model/riscv_insts_custom.sail
@@ -0,0 +1,19 @@
+union clause ast = MIN3 : (regidx, regidx, regidx)
+
+/* R-Type custom-0 encoding:
+   funct7 = 0000000
+   funct3 = 000
+   opcode = 0001011
+   https://docs.riscv.org/reference/isa/unpriv/rv-32-64g.html
+ */
+mapping clause encdec = MIN3(rs2, rs1, rd) <-> 0b0000000 @ rs2 @ rs1 @ 0b000 @ rd @ 0b0001011
+
+function clause execute (MIN3(rs2, rs1, rd)) = {
+    let rs1_val = X(rs1);
+    let rs2_val = X(rs2);
+    let smaller_val : xlenbits = if rs1_val <_s rs2_val then rs1_val else rs2_val;
+    X(rd) = smaller_val;
+    RETIRE_SUCCESS
+}
+
+mapping clause assembly = MIN3(rs2, rs1, rd) <-> "min3" ^ spc() ^ reg_name(rd) ^ sep() ^ reg_name(rs1) ^ sep() ^ reg_name(rs2)
\ No newline at end of file
