diff --git a/Makefile b/Makefile
index c0b1115..432f9b0 100644
--- a/Makefile
+++ b/Makefile
@@ -112,6 +112,7 @@ endif
 		model/riscv_insts_zicbom.sail \
 		model/riscv_insts_zicboz.sail \
 		model/riscv_jalr_seq.sail \
+		model/riscv_insts_custom.sail \
 		model/riscv_insts_end.sail \
 		model/riscv_csr_end.sail \
 		model/riscv_step_common.sail \
@@ -205,6 +206,7 @@ endif
 		model/riscv_insts_zicbom.sail \
 		model/riscv_insts_zicboz.sail \
 		model/riscv_jalr_seq.sail \
+		model/riscv_insts_custom.sail \
 		model/riscv_insts_end.sail \
 		model/riscv_csr_end.sail \
 		model/riscv_step_common.sail \
